securedrop_client/export/disk.py

```diff
@@ -0,0 +1,315 @@
+        self._export_service.connect_signals(disk_check_requested=self._poller.polling.entered)
+
+        self._poller.wait_on(self._export_service.luks_encrypted_disk_found)
+        self._poller.wait_on(self._export_service.luks_encrypted_disk_not_found)
+
+        self._poller.start_on(self.client_started_watching)
+        self._poller.pause_on(self.last_client_stopped_watching)
+
+        self._export_service.luks_encrypted_disk_not_found.connect(
+            lambda error: self._on_luks_encrypted_disk_not_found(error)
+        )
+        self._export_service.export_failed.connect(self._on_export_failed)
+        self._export_service.export_succeeded.connect(self._on_export_succeeded)
+
+    @property
+    def status(self) -> Status:
+        return self._cache.status
+
+    @property
+    def last_error(self) -> Optional[CLIError]:
+        return self._last_error  # FIXME Returning the CLIError type is an abstraction leak.
+
+    @pyqtSlot()
+    def watch(self) -> None:
+        self._connected_clients += 1
+        self.client_started_watching.emit()
+
+    @pyqtSlot()
+    def stop_watching(self) -> None:
+        self._connected_clients -= 1
+        if self._connected_clients < 1:
+            self.last_client_stopped_watching.emit()
+
+    def export_on(self, signal: pyqtBoundSignal) -> None:
+        """Allow to export files, in a thread-safe manner."""
+        self._export_service.connect_signals(export_requested=signal)
+
+    def check_status_once_on(self, signal: pyqtBoundSignal) -> None:
+        warnings.warn(
+            "check_status_once_on must not be used for new features, use the connect method instead",  # noqa: E501
+            DeprecationWarning,
+            stacklevel=2,
+        )
+
+        self._export_service.connect_signals(disk_check_requested=signal)
+
+    @pyqtSlot()
+    def _on_luks_encrypted_disk_not_found(self, error: CLIError) -> None:
+        self._last_error = error
+
+    @pyqtSlot()
+    def _on_export_succeeded(self) -> None:
+        self.export_done.emit()
+
+    @pyqtSlot(object)
+    def _on_export_failed(self, error: CLIError) -> None:
+        self._last_error = error
+        self.export_failed.emit()
+
+
+class _StatusCache(QStateMachine):
+    """A cache that holds the information available about the status of the export disk.
+
+    Paste the following state chart in https://mermaid.live for
+    a visual representation of the behavior implemented by this class!
+    stateDiagram-v2
+      [*] --> unknown
+      unknown --> luks_encrypted: export_service.luks_encrypted_disk_found
+      unknown --> unreachable: export_service.luks_encrypted_disk_not_found
+
+      luks_encrypted --> luks_unreachable: export_service.luks_encrypted_disk_not_found
+      luks_encrypted --> unknown: registered_clearing_signal_emitted
+
+      unreachable --> luks_encrypted: export_service.luks_encrypted_disk_found
+      unreachable --> unknown: registered_clearing_signal_emitted
+    """
```

Comment:
Thanks for the hint to mermaid.live, very nice! :)

I have a few questions about these states:
- I initially expected the possible states to roughly map onto the values that `sd-export` returned, or maybe to include a further comment in the docstring on why that's not so. (So for example, if this state machine is to track whether a disk is "attached," "attached+writable", or "none of the above", as opposed to the granularity that is offered by sd export, then maybe we can specify this in the docstring.)
- Relatedly, if I'm understanding correctly, maybe we can rename `luks_encrypted*`/`luks_reachable_*` to more generic terms, to be clear that this is about whether the disk is reachable or not, and not about whether the type of encryption is supported or not. (This will also avoid having to confusion when VeracCrypt support is introduced)

Reply:
I made the state machine very specifically adjusted to what we support _today_ intentionally, because that will be a readable diff to made when we add VeraCrypt support / stop using the legacy `sd-devices` API.

My reasoning was that currently either you have a LUKS-encrypted device that's reachable, or you're out of luck. And I wanted the state machine to reflect the fact that those two aspects are currently coupled. We can talk bout this. :+1:

Reply:
On your first bullet point: the state machine reflects what the GUI needs, not what `sd-devices` offers. That's intentional. Now, having said that, the conversation is open on how granular we want the GUI to be!

Reply:
Done. Renamed to `Disk.StatusReachable` by contract with `Disk.StatusUnreachable`.